### ISSUE TITLE
sharded fft unnecessarily loads entire array

### ISSUE THREAD

**Issue by ItayKishon-Remondo** [google/jax#15680]
### Description

A 2D array is sharded over several devices, slicing in a certain axis (0). When performing FFT over the other axis (1), the entire array is intermediately copied to each device, although it already contains the local slice required for the operation.

```
# Use 4 CPUs, for simple demo
import os
os.environ["XLA_FLAGS"] = "--xla_force_host_platform_device_count=4"
from jax.config import config
config.update('jax_platform_name', 'cpu')

import jax
import jax.numpy as jnp
from jax.experimental import mesh_utils
from jax.sharding import PositionalSharding

sharding = PositionalSharding(mesh_utils.create_device_mesh( (jax.device_count(),))).reshape([jax.device_count(),-1])
fft_input = jax.device_put(jnp.ones((1000,1000)), sharding)

jit_fft = jax.jit(lambda a: jnp.fft.fft(a=a, n=4096, axis=1), in_shardings=sharding, out_shardings=sharding)
jit_fft_lowered  = jit_fft.lower(fft_input)
jit_fft_compiled = jit_fft_lowered.compile()

print(jit_fft_compiled.as_text())
```

Looking at the main function of the compiled output, it seems that although the main function inputs & outputs only a single slice's shape (as expected), there is an intermediate stage which collects the slices from all devices:

```
ENTRY %main.20_spmd (param: f32[250,1000]) -> c64[250,4096] {
  %constant.1 = u8[136]{0} constant({...})
  %partition-id.1 = u32[] partition-id()
  %replica-id = u32[] replica-id()
  %param = f32[250,1000]{1,0} parameter(0), sharding={devices=[4,1]0,1,2,3}
  %call.1 = c64[1000,4096]{1,0} call(u32[] %partition-id.1, u32[] %replica-id, f32[250,1000]{1,0} %param), to_apply=%parallel_fusion.1
  %all-reduce = c64[1000,4096]{1,0} all-reduce(c64[1000,4096]{1,0} %call.1), channel_id=1, replica_groups={{0,1,2,3}}, use_global_device_ids=true, to_apply=%add
  %custom-call.1 = c64[1000,4096]{1,0} custom-call(u8[136]{0} %constant.1, c64[1000,4096]{1,0} %all-reduce), custom_call_target="ducc_fft", operand_layout_constraints={u8[136]{0}, c64[1000,4096]{1,0}}, api_version=API_VERSION_STATUS_RETURNING, metadata={op_name="jit(<lambda>)/jit(main)/jit(fft)/fft[fft_type=FftType.FFT fft_lengths=(4096,)]" source_file="/home/itayk/git/Sandbox/ItayK/Playground/jax/multi_GPU__Apr23/jit_distributed_fft__minimal_A.py" source_line=19}
  ROOT %call = c64[250,4096]{1,0} call(c64[1000,4096]{1,0} %custom-call.1, u32[] %partition-id.1), to_apply=%parallel_fusion
}
```

You can also find a [detailed example](https://gist.github.com/ItayKishon-Remondo/d04de8ccc5f19a683538235380ffc5c3) (using 4 GPUs), as well as a larger [context for the problem](https://github.com/google/jax/discussions/13842#discussioncomment-5637405).

Am I operating this in a wrong way?

Thank you.

### What jax/jaxlib version are you using?

jax 0.4.8, jaxlib 0.4.7+cuda12.cudnn88

### Which accelerator(s) are you using?

4 x nvidia T4 / CPU (both versions)

### Additional system info

Python 3.9.12, Ubuntu 22.04 LTS (both native & WSL)

**Comment by mehdiataei** [comment]
Laziness was removed from array creation in device_put. I think that's why this is happening.

See: https://github.com/google/jax/pull/14953

**Comment by ItayKishon-Remondo** [comment]
It seems this is the expected behavior: "_..., by default, `jax.numpy.fft.fft` does not understand sharding and will replicate the global array on all GPUs before computing FFTs._" (see below for source).

One of the topics in last month's Nvidia GTC, during [What's New in JAX session](https://www.nvidia.com/en-us/on-demand/session/gtcspring23-s51956/), was large scale FFT (@33:15). @nvlcambier has presented this topic (_thank you; that was the most informative single minute I've watched in a long time_), showing several different solutions for the problem, using either native JAX (with `custom_partitioning`) or via nvidia's `cuFFTMp` library:

https://github.com/NVIDIA/CUDALibrarySamples/tree/master/cuFFTMp/JAX_FFT

_BTW: I've also solved it natively in JAX, by using the fairly recent `shard_map`. Nonetheless, I think Leopold's solutions are much better._

For the Jax admins:
Prior to closing this issue, I would like to make several suggestions; should they be opened under a different issue(s)?
1. Add a note regarding this limitation (and proposed solution) in the [jax.numpy.fft.fft](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.fft.fft.html) doc.
2. It seems that the `cufft` library [does support multi-GPU, though only locally](https://docs.nvidia.com/cuda/cufft/index.html#multiple-gpu-data-organization-for-single-2d-and-3d-transforms). Should this be added natively to Jax?
3. Furthermore, is it worthwhile to consider native support for [cuFFTMp](https://docs.nvidia.com/hpc-sdk/cufftmp/index.html) as well?

Cheers

**Comment by nvlcambier** [comment]
In case that's helpful, there is some example code with `@custom_partitioning` and `jax.numpy.fft.fft` here https://jax.readthedocs.io/en/latest/jax.experimental.custom_partitioning.html in the "Example" section.

**Comment by mjsML** [comment]
It would be helpful if you can send PRs :)

**Comment by dkn16** [comment]
I found that not only `fft`, `jnp.linalg.cholesky()` has similar problem. Thanks for the `shard_map` solution, it fits my need too.

**Comment by francois-rozet** [comment]
What I find annoying about this bug, is that using `jax.vmap(jnp.fft.fft)(x)` instead of `jnp.fft.fft(x, axis=1)` does not fix it. I would have expected that a vmapped operation would never attempt to "access" (including copying from other devices) another element of the array along the mapped axis.

Would it be sensible to modify `jax.vmap` (or create a new decorator) to tell the compiler (similar to `custom_partitioning`) that a local slice along an axis is enough to perform the computation?

**Comment by clemisch** [comment]
> Would it be sensible to modify jax.vmap (or create a new decorator) to tell the compiler that a local slice along an axis is enough to perform the computation?

Isn't that always the case, by definition?

**Comment by francois-rozet** [comment]
> Isn't that always the case, by definition?

Apparently no. My guess is that the compiler optimizes the `jax.vmap` away, if the operation is already "batchable".

**Comment by jakevdp** [comment]
Just to be clear, the compiler knows nothing of `vmap`. `vmap` is entirely resolved during tracing in the Python layer, before the IR is generated and passed to the compiler.
